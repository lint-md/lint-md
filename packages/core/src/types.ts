import { MarkdownNode } from '@lint-md/parser/lib/tmp';
import { createFixer } from './utils/fixer';
import { createRuleManager } from './utils/rule-manager';

export interface MarkdownNodePosition {
  /**
   * 所在行（索引从 1 开始）
   */
  line: number;

  /**
   * 所在列（索引从 1 开始）
   */
  column: number;
}

export interface Fix {
  range: TextRange;
  text: string;
}

export interface ReportOption {
  message: string;
  loc: {
    start: MarkdownNodePosition;
    end: MarkdownNodePosition;
  };
  fix?: (fixer: ReturnType<typeof createFixer>) => Fix;
}

export interface LintMdRule {
  /**
   * 选择器初始化回调
   */
  create: (
    context: ReturnType<
      ReturnType<typeof createRuleManager>['createRuleContext']
    >
  ) => Record<string, (node: MarkdownNode) => void>;

  /**
   * rule 的一些基本信息，后续有需要再补充
   */
  meta?: Record<any, any>;
}

export interface LintMdRuleConfig {
  rule: LintMdRule;
  options?: Record<string, any>;
  fileName?: string;
}

// 节点队列
export interface NodeQueue {
  node: MarkdownNode;
  isEntering: boolean;
}

// 遍历器的相关选项
export interface TraverserOptions {
  // 在节点进入时做些什么
  onEnter?: (node: MarkdownNode, parent: MarkdownNode) => void;

  // 在节点退出时做些什么
  onLeave?: (node: MarkdownNode, parent: MarkdownNode) => void;
}

export type TextRange = number[];
