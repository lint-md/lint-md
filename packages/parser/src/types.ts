import { Parent } from 'unist';

export type MarkdownNode = Omit<Parent, 'children'> & {
  children?: MarkdownNode[];
};
