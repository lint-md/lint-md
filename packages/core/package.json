{
  "name": "@lint-md/core",
  "version": "0.2.1",
  "description": "Core of lint-md which used to lint your markdown file for Chinese.",
  "main": "lib/index.js",
  "module": "esm/index.js",
  "contributors": [
    "hustcc <i@hust.cc>",
    "yuzhanglong <loveyzl1123@gmail.com>"
  ],
  "scripts": {
    "lib:cjs": "tsc -p tsconfig.json --target ES5 --module commonjs --outDir lib",
    "lib:esm": "tsc -p tsconfig.json --target ES5 --module ESNext --outDir esm",
    "build": "run-p lib:*",
    "clean": "rimraf lib esm"
  },
  "files": [
    "esm",
    "lib",
    "src"
  ],
  "keywords": [
    "lint",
    "markdown",
    "lint-md",
    "ast-plugin"
  ],
  "dependencies": {
    "lodash": "^4.17.21",
    "remark-parse": "^9",
    "unified": "^9"
  },
  "author": "hustcc",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/lint-md/lint-md"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/lint-md.git"
  },
  "homepage": "https://github.com/lint-md#readme"
}
