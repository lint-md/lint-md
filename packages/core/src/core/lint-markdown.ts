import { parseMd } from '@lint-md/parser/lib/tmp';
import { LintMdRuleConfig, NodeQueue } from '../types';
import { createEmitter } from '../utils/emitter';
import { createTraverser } from '../utils/traverser';
import { createRuleManager } from '../utils/rule-manager';

/**
 * 基于各种 rules 对 Markdown 文本进行校验
 *
 * @date 2021-12-12 21:48:21
 */
export const lintMarkdown = (
  markdown: string,
  allRuleConfigs: LintMdRuleConfig[]
) => {
  // 将 markdown 转换成 ast
  const ast = parseMd(markdown);

  // 节点队列，遍历到的节点都会被推入这里
  const nodeQueue: NodeQueue[] = [];

  // 全局规则管理器
  const ruleManager = createRuleManager();

  // 初始化遍历器，对于每一个节点的进入和退出，都将其推入到上面的 nodeQueue 队列中，供后续处理
  const traverser = createTraverser({
    onEnter: (node) => {
      nodeQueue.push({
        isEntering: true,
        node: node,
      });

      nodeQueue.push({
        isEntering: false,
        node: node,
      });
    },
  });

  const emitter = createEmitter();

  // 遍历所有的 rules，并拿到它们的选择器，为每一个选择器订阅相关事件
  for (const { rule, options } of allRuleConfigs) {
    const ruleContext = ruleManager.createRuleContext(options);
    const ruleSelectors = rule.create(ruleContext);
    for (const selector of Object.keys(ruleSelectors)) {
      emitter.on(selector, ruleSelectors[selector]);
    }
  }

  // 递归地遍历 ast
  traverser.traverse(ast, null);

  // 遍历节点队列，执行对应的选择器
  for (const nodeQueueItem of nodeQueue) {
    const { node, isEntering } = nodeQueueItem;

    try {
      if (isEntering && node.type) {
        emitter.emit(node.type, node);
      }
    } catch (e) {
      console.log(e);
    }
  }

  return {
    ruleManager: ruleManager,
  };
};
