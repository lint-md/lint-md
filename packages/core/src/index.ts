export { lintMarkdown } from './core/lint-markdown';
