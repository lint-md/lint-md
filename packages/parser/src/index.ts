export { parseMd } from './parse-md'
