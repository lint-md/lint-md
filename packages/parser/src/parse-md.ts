import { unified } from 'unified'
import remarkParse from 'remark-parse'

/**
 * 将 Markdown 解析成 ast
 *
 * TODO: 完善返回值的类型定义
 *
 * @param {string} md Markdown 文本
 * @returns {any} md ast 结构
 * @author YuZhanglong <loveyzl1123@gmail.com>
 */
export const parseMd = (md: string): any => {
  return unified().use(remarkParse).parse(md)
}
