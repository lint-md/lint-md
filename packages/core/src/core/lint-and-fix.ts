import { Fix, LintMdRuleConfig } from '../types';
import { applyFix } from '../utils/apply-fix';
import { MAX_LINT_AND_FIX_CALL_TIMES } from '../common/constant';
import { lintMarkdown } from './lint-markdown';

/**
 * 核心方法，对某个 Markdown 文本进行 lint 或者 fix
 *
 * @date 2021-12-14 17:16:12
 */
export const lintAndFix = (markdown: string, rules: LintMdRuleConfig[], isFixMode: boolean) => {
  let lintTimes = 0;
  let lintResult = lintMarkdown(markdown, rules);

  let fixedResult: { result: string, notAppliedFixes: Fix[] } = null;

  while (isFixMode && lintTimes <= MAX_LINT_AND_FIX_CALL_TIMES) {
    lintTimes += 1;
    fixedResult = applyFix(markdown, lintResult.ruleManager.getAllFixes());
    if (!fixedResult.notAppliedFixes.length) {
      break;
    }
    lintResult = lintMarkdown(markdown, rules);
  }

  return {
    lintResult: lintResult,
    fixedResult: fixedResult
  };
};
