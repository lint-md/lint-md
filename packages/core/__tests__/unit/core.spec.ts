import { isFunction } from 'lodash';
import { lintMarkdown } from '../../src';
import noEmptyCode from '../../src/rules/no-empty-code';
import { lintAndFix } from '../../src/core/lint-and-fix';

describe('test core methods for lint-markdown', () => {
  test('test lintMarkdown() to lint source', () => {
    const lintResult = lintMarkdown(`# Hello

Some **importance**, and \`code\`.

\`\`\`javascript

\`\`\``, [
      {
        rule: noEmptyCode
      }
    ]);

    expect(lintResult.ruleManager.getReportData().length).toStrictEqual(1);
    const res = lintResult.ruleManager.getReportData().pop();
    expect(isFunction(res.fix)).toBeTruthy();
    expect(res.message).toStrictEqual('[lint-md] 代码块内容不能为空，请删除空的代码块，或者填充代码内容');
  });

  test('test lintAndFix() to lint or fix markdown source', () => {
    const res = lintAndFix(`# Hello

Some **importance**, and \`code\`.

\`\`\`javascript

\`\`\``, [
      {
        rule: noEmptyCode
      }
    ], true);

    expect(res.fixedResult.notAppliedFixes).toStrictEqual([]);
    expect(res.fixedResult.result).toMatchSnapshot();
  });
});
