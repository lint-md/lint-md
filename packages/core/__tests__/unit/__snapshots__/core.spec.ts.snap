// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test core methods for lint-markdown test lintAndFix() to lint or fix markdown source 1`] = `
"# Hello

Some **importance**, and \`code\`.

"
`;
